url1d5e9d990f_Telsis_jOCP_tgz-pJ8-com_telsis_jocp_sampleApp_BasicOCPAppJ8/src/com/telsis/jocp/OCPSystemManager.java

```diff
@@ -0,0 +1,1344 @@
+                InetAddress preferredUnit = links.get(0).preferredSCP;
+                InetAddress secondaryUnit = links.get(0).secondarySCP;
+
+                if (preferredUnit != null && secondaryUnit != null) {
+                    unitsMatch = true;
+
+                    for (Link link : links) {
+                        if (!preferredUnit.equals(link.preferredSCP)
+                                && !secondaryUnit.equals(link.secondarySCP)) {
+                            unitsMatch = false;
+                            break;
+                        }
+                        if (preferredUnit.equals(link.linkManager
+                                .getRemoteAddress())) {
+                            newPreferred = link;
+                        }
+                    }
+
+                    // Note that it is possible to get here without a preferred
+                    // link even if all units agree, if the link they agree on
+                    // is not one of ours. In that case we don't have a
+                    // preferred link.
+                }
+
+                if (!unitsMatch) {
+                    newPreferred = null;
+                }
+            }
+            break;
+
+        case MASTER_SLAVE:
+            for (Link link : links) {
+                if (link.linkState == LinkStates.MASTER) {
+                    newPreferred = link;
+                }
+            }
+            break;
+
+        default:
+            newPreferred = null;
+            break;
+        }
+
+        if (newPreferred != preferredLink) {
+            log.info("recalcPreferredUnit: changed from " + preferredLink
+                    + " to " + newPreferred);
+            preferredLink = newPreferred;
+        }
+    }
+
+    /**
+     * Select an OCP link to use for a new call. The link chosen depends on the
+     * current mode:
+     * <p/>
+     * <b>Master/slave:</b> if the master can take calls, then the master is
+     * returned. Otherwise null is returned.
+     * <p/>
+     * <b>Loadsharing:</b> if the connected servers have selected a preferred
+     * server, then this is returned. Otherwise the server with the fewest
+     * active calls is returned.
+     * <p/>
+     * Additionally, any links that are specified in triedLinks will not be
+     * returned by this function. If the above process would select one of these
+     * then the next available link will be returned instead. This function will
+     * also not return any links that are considered to be suspect or are not
+     * currently capable of taking new calls.
+     *
+     * @param triedLinks
+     *            A list of links that have already been tried and so may not be
+     *            returned by this function.
+     * @return The OCP link to try next. If null is returned then there are no
+     *         more available links.
+     */
+    @Override
+    public final OCPLink getLink(final List<OCPLink> triedLinks) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 14. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.